### ISSUE TITLE
The initial value of the file register for DWARF5 line number information state machine is incorrect

### ISSUE THREAD

**Issue by freejack811** [davea42/libdwarf-code#137]
Section 6.2.2 and Table 6.4(page153) of the DWARF5 spec. clearly states that the initial value of the file register should be set to 1

So, the following lines of code should be removed:
https://github.com/davea42/libdwarf-code/blob/602111edf64e1c92e7784da695188a508517e0d4/src/lib/libdwarf/dwarf_line.c#L82-L92

I'm guessing the comment in the code above is a misunderstanding of the following text(page158:1-8) in spec.:

_The line number program references file names in this sequence beginning
with 0, and uses those numbers instead of file names in the line number
program that follows.
Prior to DWARF Version 5, the current compilation file name was not represented in
the file_names field. In DWARF Version 5, the current compilation file name is
explicitly present and has index 0. This is needed to support the common practice of
stripping all but the line number sections (.debug_line and debug_line_str)
from an executable._ 

The above content don't imply that the initial value of the file register is 0. 

I ran into a problem when compiling a program with gcc on my machine.  When I use dwarfdump to dump the program I get the following information：

_.debug_line: line number info for a single cu
Source lines (from CU-DIE at .debug_info offset 0x0000000c):
            NS new statement, BB new basic block, ET end of text sequence
            PE prologue end, EB epilogue begin
            IS=val ISA number, DI=val discriminator value
<pc>        [lno,col] NS BB ET PE EB IS= DI= uri: "filepath"
**0x00015452  [  66, 5] NS uri: "/home/osum/libst/main.cc"
0x00015462  [  66,50] NS
0x00015473  [  66,54] NS
0x00015476  [  66,54] NS ET**
0x00015476  [ 124, 1] NS uri: "/home/osum/libst/include/st_file.h"
0x0001548a  [ 125, 5] NS
0x00015491  [ 126,21] NS
0x000154a1  [ 127,24] NS
0x000154be  [ 128, 5] NS
0x000154c5  [ 129,15] NS
0x000154d1  [ 130,12] NS
0x000154d5  [ 131, 1] NS
0x000154d7  [ 131, 1] NS ET
0x000154d8  [  28, 5] NS uri: "/home/osum/libst/main.cc"_
...

The bold text above indicates that the code is on line 66 of main.cc, but it's actually on line 66 in another header file, where I define an inline function.

$ gcc --version
gcc (Ubuntu 11.2.0-19ubuntu1) 11.2.0
Copyright (C) 2021 Free Software Foundation, Inc.

**Comment by davea42** [comment]
It's possible there is something interesting about your source files that results in the
dwarfdump output.   Could you create a non-proprietary not-too-large object file
that shows this kind of problem and get it to me somehow?

**Comment by davea42** [comment]
I created a small testcase in C and all is well.

Then I built dwarfgen (C++)  and the line table resulting
really seems very odd.

The initial entries in the file table for each CU say
something like:
0x00009600  [1087, 7] NS uri: "/home/davea/dwarf/code/src/bin/dwarfgen/irepattrtodbg.cc"
0x00009600  [1087, 7]
0x00009604  [1088, 9] NS
0x00009604  [1087, 7]
0x00009606  [1088,21]
0x00009607  [1088,21] ET
with pc  and line number the same but the uri: file name changing.

At the moment I do not understand what that means.
I find that x9600 to x9607 is a DW_TAG_subprogram
with
 DW_AT_low_pc                0x00009600

and CUs have it in rangelists as
one range:
    [ 2] DW_RLE_start_length   0x00009600 0x00000007
      [ 2] start,end             0x00009600 0x00009607

linkage name _ZNKSt5ctypeIcE8do_widenEc
In other words, this seems to be  a g++  generated function of inlined code
and any line number involved seems to be irrelevant to you or I.
My build has five COMPILE_UNITs  Every compile unit one leads off with this inlined function
and assigns a source name to it. A different source file name every time.

So I suggest you check your object file to see if you see some similar arrangement.
Let us know what you find.

**Comment by davea42** [comment]
For now I'll stop here. The compiler is showing the existence of real code but
since it's not code you wrote it is confusing.  Note attributes
 DW_AT_inline DW_AT_abstract_origin in the various instances.

**Comment by davea42** [comment]
I  guess it should be clear, now, that the default file number of 0 in regs->lr_file
is neither a misunderstanding nor an error.  It is correcting a mistake in the standard
in a way consistent with current compiler DWARF5 generation.

**Comment by freejack811** [comment]
I also think there is a certain description error in the DWARF5 standard, but only to some of the stuff in section 2.14 that you already mentioned in the source code comments. But the initial value of the file register of the state machine is very deterministic (see Table 6.4 on page 163 of the DWARF5 standard). 

Even though some places in the standard say that index 0 can refer to a compilation unit file, and some places say that index 0 refers to no file, which is indeed contradictory, the standard does not mandate the use of index 0. The gcc uses the initial value of 1 to generate debugging information, and never uses a file index value less than 1, which avoids the contradiction of the standard, so gcc does not violate the standard in any way. 

Maybe this initial value will be changed in DWARF6, or there will be more changes, but version 5 has been been released for a long time, and many implementations implement the current standard, so as far as version 5 is concerned, we should respect the current implementation. Later I'll write a simple example of how gcc works and test it out to see what clang does.

**Comment by freejack811** [comment]
Below is the code of my test program.
test.h
```
#ifndef TEST__H_
#define TEST__H_

// Deliberately left blank lines, so that the number of lines where the inline
// function is located is greater than all the lines of code in main.cc






static inline int inline_add(int a) {
    return ++a;
}


#endif
```

**Comment by freejack811** [comment]
main.cc
```#include "test.h"

#include <cstdio>

int main(int /*argc*/, char* [] /*argv[]*/) {
    int a = inline_add(1);
    printf("%d\n", a);
}
```

**Comment by freejack811** [comment]
Put the above two files into the current directory and generate a.out
#gcc -g main.cc
#dwarfdump a.out
The row information we get is as follows:
_...
.debug_line: line number info for a single cu
Source lines (from CU-DIE at .debug_info offset 0x0000000c):
.
            NS new statement, BB new basic block, ET end of text sequence
            PE prologue end, EB epilogue begin
            IS=val ISA number, DI=val discriminator value
<pc>        [lno,col] NS BB ET PE EB IS= DI= uri: "filepath"
**0x00001149  [  12,37] NS uri: "/home/sutao/stlibrary/build/main.cc"
0x00001150  [  13,14] NS
0x00001157  [  14, 1] NS**
0x00001159  [   5,45] NS
0x0000116c  [   6,23] NS
0x00001179  [   7,11] NS
0x00001192  [   8, 1] NS
0x00001199  [   8, 1] NS ET
..._
The bold text above will show that the code is in lines 13-14 of main.cc, but in fact our main.cc is only 8 lines of code, which actually refers to lines 13-14 in test.h. Therefore, the content displayed by dwarfdump is wrong. The reason is that the initial value of the file register in version 5 is set to 0 in dwarf_line.c in libdwarf that I mentioned at first. I use gdb to debug, gdb seems to use libbfd, so it appears correct:

_#gdb ./a.out
(gdb) b inline_add
Breakpoint 1 at 0x1150: file /home/sutao/stlibrary/build/test.h, line 13.
(gdb) run
Starting program: /home/sutao/stlibrary/build/a.out 
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/lib/x86_64-linux-gnu/libthread_db.so.1".

Breakpoint 1, inline_add (a=1) at /home/sutao/stlibrary/build/test.h:13
13	    return ++a;
(gdb) bt
#0  inline_add (a=1) at /home/sutao/stlibrary/build/test.h:13
#1  0x0000555555555176 in main () at main.cc:6
(gdb)_

**Comment by freejack811** [comment]
Very strange, I can't get line information with clang13.1 generation

 % clang --version
Apple clang version 13.1.6 (clang-1316.0.21.2.5)
Target: x86_64-apple-darwin21.4.0
Thread model: posix
InstalledDir: /Library/Developer/CommandLineTools/usr/bin

%clang -gdwarf-5 ./main.cc
% ./dwarfdump --version
./dwarfdump [Oct 11 2022 12:01:51 (libdwarf 0.4.2 dwarfdump 0.4.2)]
%./dwarfdump -a ./a.out.dSYM/Contents/Resources/DWARF/a.out
```
...
< 1><0x00000868>    DW_TAG_pointer_type
                      DW_AT_type                  <GOFF=0x00000306>

.debug_line: line number info for a single cu

.debug_str
name at offset 0x00000000, length    0 is ''
name at offset 0x00000001, length   48 is 'Apple clang version 13.1.6 (clang-1316.0.21.2.5)'
name at offset 0x00000032, length    7 is 'main.cc'
name at offset 0x0000003a, length   51 is '/Library/Developer/CommandLineTools/SDKs/MacOSX.sdk'
name at offset 0x0000006e, length   10 is 'MacOSX.sdk'
...
```

**Comment by freejack811** [comment]
Hi Dave, do you have clang14 and if so can you test it? I took a look and it seems that LLVM 13 doesn't support dwarf5 well enough, but I tried to install LLVM 14 without success, I have some things to work on, if you have time you can test it with the code above. Also, please forgive me for not replying to your message in time due to the time difference.

**Comment by davea42** [comment]
I created a new test and verified what freejack811 says.
The quoted code assumed, incorrectly, what gcc compilers
would do.

Clang (at least as of clang14) avoids ever  using the
default line table file number.
the next push to github has this fixed.

Anyone suspecting a file number error might find that
doing
```
dwarfdump -l test.o
dwarfdump -l -v test.o
dwarfdump -l -vv test.o
```
and comparing these (not with diff, but by looking)
might find it a helpful exercise.

**Comment by davea42** [comment]
Just pushed the correction to the default line number
and a few small  dwarfdump and documentation changes.
So I think this can be closed. If need be it can be reopened.

**Comment by davea42** [comment]
I take it no objections to closing.
